packages/resin-event-log/src/resin-event-log.js

```diff
@@ -0,0 +1,134 @@
+		return callback()
+	},
+	afterCreate: function(error, type, jsonData, applicationId, deviceId) {}
+}
+
+var ADAPTORS = [
+	require('./adaptors/ga'),
+	require('./adaptors/mixpanel')
+]
+
+module.exports = function(options) {
+	options = options || {}
+	var prefix = options.prefix,
+		debug = options.debug
+	if (!prefix) {
+		throw Error('`prefix` is required.')
+	}
+
+	var hooks = assign(
+		{},
+		DEFAULT_HOOKS,
+		pick(options, keys(DEFAULT_HOOKS))
+	)
+
+	var adaptors = ADAPTORS.map(function (adaptorFactory) {
+		return adaptorFactory(options)
+	}).filter(function (adaptor) {
+		// Skip the adaptors that did not init (due to missing config options)
+		return !!adaptor
+	})
+
+	function runForAllAdaptors(methodName, args, callback) {
+		return Promise.all(
+			adaptors.map(function (adaptor) {
+				return adaptor[methodName]
+					? adaptor[methodName].apply(adaptor, args)
+					: null
+			})
+		).asCallback(callback)
+	}
+
+	var eventLog = {
+		userId: null,
+		prefix: prefix,
+		start: function(user, callback) {
+			if (!user) {
+				return Promise.reject(new Error(
+					'user is required to start events interaction.'
+				)).asCallback(callback)
+			}
+
+			this.userId = user.id
+
+			return runForAllAdaptors('login', [ user ], callback)
+		},
+		end: function(callback) {
+			this.userId = null
+			return runForAllAdaptors('logout', [], callback)
+		},
+		create: function(type, jsonData, applicationId, deviceId, callback) {
+			var _this = this
+
+			function runBeforeHook() {
+				return Promise.fromCallback(function(callback) {
+					hooks.beforeCreate.call(_this, type, jsonData, applicationId, deviceId, callback)
+				}).catch(function (err) {
+					// discard the hook error
+					if (debug) {
+						console.warn("`beforeCreate` error", err)
+					}
+				})
+			}
+
+			function runAfterHook(err) {
+				return Promise.try(function() {
+					hooks.afterCreate.call(_this, err, type, jsonData, applicationId, deviceId)
+				}).catch(function (err) {
+					// discard the hook error
+					if (debug) {
+						console.warn("`afterCreate` error", err)
+					}
+				})
+ 			}
+
+			return runBeforeHook()
+				.then(function() {
+					return runForAllAdaptors('track', [
+						_this.prefix, type, {
+							applicationId: applicationId,
+							deviceId: deviceId,
+							jsonData: jsonData
+						}
+					])
+				}).catch(function (err) {
+					// catch the tracking error and call the hook
+					runAfterHook(err)
+					// rethrow the error to not call the hook for the second time in the next `.then`
+					throw err
+				}).then(function () {
+					return runAfterHook()
+				}).asCallback(callback)
+		}
+	}
+
+	keys(EVENTS).forEach(function (base) {
+		var events = EVENTS[base]
+		var obj = eventLog[base] = {}
+		events.forEach(function(event) {
+			obj[event] = function(jsonData, applicationId, deviceId) {
+				return eventLog.create(startCase(base + " " + event), jsonData, applicationId, deviceId)
+			}
```

Comment:
EVENTS is an object, not an array, so no forEach unless you want to use lodash _.each.
As the code is universal I try to use as few lodash methods as possible, so no zipObject.